File: Landmark-Kazemi/python/kazemi-webcam-demo.py
import os
import time
import cv2
import argparse

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Demonstration of kazemi facial landmark algorithm.')
    parser.add_argument('--face_cascade', type=str, help="Path to the cascade model file for the face detector",
                        default=os.path.join(os.path.dirname(os.path.realpath(__file__)),'models','lbpcascade_frontalface_improved.xml'))
    parser.add_argument('--kazemi_model', type=str, help="Path to the kazemi trained model file",
                        default=os.path.join(os.path.dirname(os.path.realpath(__file__)),'models','face_landmark_model.dat'))
    parser.add_argument('--from_video', type=str, help='Use this video path instead of webcam')
    parser.add_argument('--record_video', action='store_true')
    parser.add_argument('--path_to_save_video',type=str, help='Output path of video of demonstration.', default='output/face_landmarks.avi')
    parser.add_argument('--width', type=int, help='width of the window', default=960)
    parser.add_argument('--height', type=int, help='height of the window', default=720)
    parser.add_argument('--camera_id', type=int, default=0, help='ID of webcam to use')
    parser.add_argument('--verbose', action='store_true')
    args = parser.parse_args()

    video_capture = None

    if args.from_video:
        assert os.path.isfile(args.from_video)
        video_capture = cv2.VideoCapture(args.from_video)
    else:
        video_capture = cv2.VideoCapture(args.camera_id)
    video_capture.set(3, args.width)
    video_capture.set(4, args.height)

    face_cascade = cv2.CascadeClassifier(args.face_cascade)
    facemark = cv2.face.createFacemarkKazemi()
    facemark.loadModel(args.kazemi_model)

    cv2.namedWindow('Landmark_Window', cv2.WINDOW_NORMAL)
    cv2.resizeWindow('Landmark_Window', args.width, args.height)

    if args.record_video:
        record_video = cv2.VideoWriter(args.path_to_save_video, cv2.VideoWriter_fourcc('M','J','P','G'), 10,
                                        (int( video_capture.get(cv2.CAP_PROP_FRAME_WIDTH)), int( video_capture.get(cv2.CAP_PROP_FRAME_HEIGHT))))

    while(video_capture.isOpened()):
        ret, frame = video_capture.read()
        if ret :
            gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
            faces = face_cascade.detectMultiScale(gray_frame, 1.3, 5)
            if len(faces) > 0:
                start = time.time()
                status, landmarks = facemark.fit(frame, faces)
                if args.verbose:
                    print("status : ", status)
                    print("landmarks : ", landmarks)
                    print("time taken to fit : {0:.2f} ms".format(1000*(time.time() - start)))
                for f in range(len(landmarks)):
                    cv2.face.drawFacemarks(frame, landmarks[f])
            else:
                if args.verbose:
                    print("No Face Detected")
            if args.record_video:
                record_video.write(frame)
            cv2.imshow('Landmark_Window', frame)
            if cv2.waitKey(1) & 0xFF == 27:
                break
        else:
            break
    video_capture.release()
    if args.record_video:
        record_video.release()
    cv2.destroyAllWindows()

File: Landmark-Kazemi/c++/kazemi-webcam-demo.cpp
/*----------------------------------------------
 * Usage:
 * kazemi-webcam-demo <face_cascade_model> <kazemi_model>
 *
 * Example:
 * kazemi-webcam-demo models/face_cascade.xml  models/face_landmark_model.dat
 *
 *--------------------------------------------------*/

#include <stdio.h>
#include <fstream>
#include <sstream>
#include "opencv2/face.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/videoio.hpp"
#include "opencv2/objdetect.hpp"
#include "opencv2/imgproc.hpp"

#include <iostream>
#include <string>
#include <vector>
#include <ctime>

using namespace std;
using namespace cv;
using namespace cv::face;

bool myDetector(InputArray image, OutputArray faces, CascadeClassifier *face_cascade)
{
    Mat gray;
    if (image.channels() > 1)
        cvtColor(image, gray, COLOR_BGR2GRAY);
    else
        gray = image.getMat().clone();

    equalizeHist(gray, gray);

    vector<Rect> faces_;
    face_cascade->detectMultiScale(gray, faces_, 1.4, 2, CASCADE_SCALE_IMAGE, Size(30, 30));
    Mat(faces_).copyTo(faces);
    return true;
}

bool parseArguments(int argc, char** argv,
    String & cascade,
    String & kazemi_model
){
    const String keys =
        "{ @f face-cascade    |      | (required) path to the cascade model file for the face detector }"
        "{ @m model-path      |      | (required) path to the trained Kazemi model}"
        "{ help h usage ?     |      |  kazemi-webcam-demo -face-cascade  -model \n"
             " example: kazemi-webcam-demo models/face_cascade.xml  models/face_landmark_model.dat}"
    ;
    CommandLineParser parser(argc, argv, keys);
    parser.about("hello");

    if (parser.has("help")){
        parser.printMessage();
        return false;
    }

    cascade = String(parser.get<string>("face-cascade"));
    kazemi_model = String(parser.get<string>("model-path"));

    if(cascade.empty() || kazemi_model.empty()){
        cerr << "one or more required arguments are not found" << '\n';
        cout<<"face-cascade : "<<cascade.c_str()<<endl;
        cout<<"kazemi-model : "<<kazemi_model.c_str()<<endl;
        parser.printMessage();
        return false;
    }
    return true;
}


int main(int argc, char** argv )
{
    String cascade_path, kazemi_model;
    if(!parseArguments(argc, argv, cascade_path, kazemi_model))
       return -1;

    /* Create the Facemark Instance */
    CascadeClassifier face_cascade;
    face_cascade.load(cascade_path);
    FacemarkKazemi::Params params;
    Ptr<FacemarkKazemi> facemark = FacemarkKazemi::create(params);
    facemark->setFaceDetector((FN_FaceDetector)myDetector, &face_cascade);
    facemark->loadModel(kazemi_model);

    /* Set up webcam for video capture */
    VideoCapture cam(0);

    /* Initialise the video writer object */
    VideoWriter video("webcam_landmarks.avi",VideoWriter::fourcc('M','J','P','G'),10, Size(cam.get(cv::CAP_PROP_FRAME_WIDTH),cam.get(cv::CAP_PROP_FRAME_HEIGHT)),true);
    if ( !video.isOpened() ) //if VideoWriter did not initliazr successfully, exit the program
    {
        cout << "ERROR: Failed to write the video" << endl;
        return -1;
    }

    Mat frame, img;
    String text;
    char buff[255];
    double fittime;
    int nfaces;
    vector<Rect> faces,faces_scaled;
    vector<vector<Point2f> > landmarks;
    namedWindow("Landmark Image", 1);

    while(cam.read(frame)){
        faces.clear();
        landmarks.clear();
        faces_scaled.clear();

        double __time__ = (double)getTickCount();
        float scale = (float)(600.0/frame.cols);
        resize(frame, img, Size((int)(frame.cols*scale), (int)(frame.rows*scale)), 0, 0, INTER_LINEAR_EXACT);
        facemark->getFaces(img,faces);

        for(int j=0;j<(int)faces.size();j++){
            faces_scaled.push_back(Rect(
                (int)(faces[j].x/scale),
                (int)(faces[j].y/scale),
                (int)(faces[j].width/scale),
                (int)(faces[j].height/scale)));
        }
        faces = faces_scaled;
        fittime=0;
        nfaces = (int)faces.size();
        if(faces.size() > 0){
            double newtime = (double)getTickCount();
            facemark->fit(frame, faces, landmarks);
            fittime = ((getTickCount() - newtime)/getTickFrequency());
            for(int j=0;j<(int)faces.size();j++){
                landmarks[j] = Mat(Mat(landmarks[j]));
                drawFacemarks(frame, landmarks[j], Scalar(0,0,255));
                rectangle(frame, faces[j], Scalar(255,0,255));
            }
        }
        double fps = (getTickFrequency()/(getTickCount() - __time__));
        sprintf(buff, "faces: %i %03.2f fps, fit:%03.0f ms",nfaces,fps,fittime*1000);
        text = buff;
        putText(frame, text, Point(20,40), FONT_HERSHEY_PLAIN , 2.0,Scalar::all(255), 2, 8);

        imshow("Landmark Image", frame);
        video.write(frame);
        if (waitKey(1) == 27)
          break;    
    }
}
